### PR TITLE
chore: add Grit patterns crate

### PR DESCRIPTION
**Pull request by arendjr**
## Summary

This PR adds more plumbing for implementing the `biome search` command. It does two things:

* Add a `biome_grit_patterns` crate that can link to the real Grit engine (Marzano is their codename). This is behind a disabled feature flag, since it pulls in a *lot* of dependencies, including TreeSitter and the grammars for all their languages.
  * Note: I am working in the Grit repository to try to eliminate their TreeSitter dependencies from the core crate.
* Update the workspace protocol to showcase how the workspace server should parse patterns and generate pattern IDs. The actual implementations are all dummies still.

~~Note: The `Cargo.lock` has a pretty huge diff because of all the transitive Grit dependencies.~~ **Edit:** Fixed by uncommenting the `marzano-core` dependency for now.

## Test Plan

Everything should stay green.

### CODE REVIEW

**Comment by arendjr** [comment]
I've commented out the `marzano-core` dependency for now, so the `Cargo.lock` doesn't include a huge amount of unused dependencies. If the tests are green, it should be safe to merge now.

File: crates/biome_parser/src/diagnostic.rs [hunk 1]
```diff
@@ -18,8 +19,9 @@ use std::cmp::Ordering;
 ///
 /// These information **are printed in this exact order**.
 ///
-#[derive(Debug, Diagnostic, Clone)]
+#[derive(Clone, Debug, Deserialize, Diagnostic, Serialize)]
```
Comment (ematipico) [inline crates/biome_parser/src/diagnostic.rs:22]: These shouldn't be deserialisable/serialisable. We have the `diagnostics::serde` module for that. Can you revert the changes *inside this file*?

File: crates/biome_service/src/diagnostics.rs [hunk 2]
```diff
@@ -56,6 +57,10 @@ pub enum WorkspaceError {
     Vcs(VcsDiagnostic),
     /// Diagnostic raised when a file is protected
     ProtectedFile(ProtectedFile),
+    /// An invalid patterns was given
+    ParsePatternError(ParseError),
+    /// No pattern with the given ID
+    InvalidPattern(InvalidPattern),
```
Comment (ematipico) [inline crates/biome_service/src/diagnostics.rs:63]: A suggestion: have a `SearchError` instead and then an `enum SearchError` that will contain these two variants. 

The search expresses the error better because that's what the feature is about. The "pattern" thing is a detail.

File: crates/biome_parser/src/diagnostic.rs [hunk 3]
```diff
@@ -32,22 +34,23 @@ pub struct ParseDiagnostic {
...
+#[derive(Clone, Debug, Default, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
 struct ParserAdvice {
     advice_list: Vec<ParserAdviceKind>,
 }
 
 /// The structure of the advice. A message that gives details, a possible range so
 /// the diagnostic is able to highlight the part of the code we want to explain.
-#[derive(Debug, Clone)]
+#[derive(Clone, Debug, Deserialize, Serialize)]
```
Comment (ematipico) [inline crates/biome_parser/src/diagnostic.rs:45]: These should not be serializable/deserialisable. When and if we require them to be serialisable, we should use `diagnostics::serde`. Here is an example:

https://github.com/biomejs/biome/blob/f6fd1ef7130e7bcd1466aa78ae7295fc5fdd5d58/crates/biome_service/src/file_handlers/json.rs#L350-L362

**Review by ematipico — COMMENTED** [review]
Can you be more descriptive about this PR? Here are some questions that I think we should answer:
- what's `marzano` and what is meant for?
- why do we need this marzano crate?
- why don't we use our grit parser?
- provide some example or use case of how this would work in the future

You have to consider that you have most of the knowledge, but nobody here knows what you're doing, so you'll have to help us understanding what's the ultimate goal, otherwise we just review coding style without having an architecture in mind.

**Comment by arendjr** [comment]
@ematipico  I’ll write a progress report with an outline of the technical direction!

**Comment by arendjr** [comment]
So with the [progress update](https://github.com/biomejs/biome/discussions/2286) out there for wider context, I can also answer your specific questions:

> what's `marzano` and what is meant for?

It's the codename for the Grit engine used by the Grit.io folks. It's part of their [open-source repository](https://github.com/getgrit/gritql/).

> why do we need this marzano crate?

I actually don't think we should be using Marzano in our end version. Instead, I want to extract a more focused crate out of Marzano that will not have any TreeSitter dependencies anymore. But until we have *that* crate, I just wanted to use Marzano for some quick 'n dirty experimentation.

> why don't we use our grit parser?

The intent is absolute to use our Grit parser, but the conversion from our CST to Grit "programs" is yet to be implemented. Until that's done, Marzano might already give us something workable. But again, that's just experimental.

> provide some example or use case of how this would work in the future

Do you mean for the `biome search` command? I see that mainly as an alternative to `git grep` except it would use GritQL queries instead of regular expressions. I'm also still interested in having the VS Code extension use the same workspace commands to allow searching for GritQL queries from the IDE.

**Review by ematipico — APPROVED** [review]
